### PR TITLE
feat: export promql service in server 

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add a promql service to serve PromQL queries. Currently only `query` and `range_query` are defined, and only `range_query` returns mock results. 

This PR also includes some minor changes in a separated commit 26d0918dd70cf3147b3bd8f8f4789733eba22701

For the name of this new service, I would prefer to rename it to `prometheus` as it's a subset of [prometheus HTTP API](https://prometheus.io/docs/prometheus/latest/querying/api/), and can be extended to implement other methods that are not included in this PR. So I propose to rename the current `prometheus` service to `prom_store`, which implements Prometheus' remote storage APIs, and rename the new service from `promql` to `prometheus`. What do you think about this @killme2008 

The code structure may need some adjustments. Specificly for `PromqlHandler` and `PromqlServer`. Do I need to put them under the HTTP server or other places? cc @MichaelScofield PTAL at this and looking forward to hearing your advice. 

An important missing part of this functionality is the codec of query results. We are using table model all around, it looks impossible to return multi-value columns in prometheus' format. Maybe we need to introduce some hack like built-in tag `__value__`. And beside this, converting should also consider the semantic type in the result's schema. This needs to change the planner and maybe the executor, so I plan to implement codec in the follow-up PR(s).

I'm also wondering if it's beneficial to provide other ways to perform promql queries, like via SDK or special SQL.

Things left to do in this PR:
- [ ] split `PromqlServer` into serval sub-modules.
- [ ] add `test-integration` case for this new service.

## Some screenshots
- data
  ```sql
  mysql> desc table demo;
  +-----------+----------------------+------+---------+---------------+
  | Field     | Type                 | Null | Default | Semantic Type |
  +-----------+----------------------+------+---------+---------------+
  | timestamp | TimestampMillisecond | NO   |         | TIME INDEX    |
  | host      | String               | YES  |         | PRIMARY KEY   |
  | value     | Float64              | YES  |         | VALUE         |
  +-----------+----------------------+------+---------+---------------+
  mysql> select * from demo;
  +-------+---------+---------------------+
  | host  | value   | timestamp           |
  +-------+---------+---------------------+
  | host1 |    66.6 | 1970-01-01 00:00:00 |
  | host1 |    66.6 | 1970-01-01 00:00:02 |
  | host1 |    66.6 | 1970-01-01 00:00:05 |
  | host1 |    43.1 | 1970-01-01 00:00:07 |
  | host1 |    19.1 | 1970-01-01 00:00:09 |
  | host1 |    99.1 | 1970-01-01 00:00:10 |
  | host1 |   999.9 | 1970-01-01 00:00:21 |
  | host1 |    31.9 | 1970-01-01 00:00:22 |
  | host1 |    95.4 | 1970-01-01 00:00:32 |
  | host1 | 12423.1 | 1970-01-01 00:00:49 |
  | host1 |       0 | 1970-01-01 00:01:20 |
  | host1 |      49 | 1970-01-01 00:01:39 |
  +-------+---------+---------------------+
  ```
- query result:
  ```
  2023-02-01T09:06:07.477917Z  INFO servers::promql: row: [Timestamp(Timestamp { value: 0, unit: Millisecond }), Float64(OrderedFloat(67.0))]
  2023-02-01T09:06:07.477960Z  INFO servers::promql: row: [Timestamp(Timestamp { value: 5000, unit: Millisecond }), Float64(OrderedFloat(67.0))]
  2023-02-01T09:06:07.477988Z  INFO servers::promql: row: [Timestamp(Timestamp { value: 10000, unit: Millisecond }), Float64(OrderedFloat(100.0))]
  2023-02-01T09:06:07.478016Z  INFO servers::promql: row: [Timestamp(Timestamp { value: 50000, unit: Millisecond }), Float64(OrderedFloat(12424.0))]
  2023-02-01T09:06:07.478044Z  INFO servers::promql: row: [Timestamp(Timestamp { value: 80000, unit: Millisecond }), Float64(OrderedFloat(0.0))]
  2023-02-01T09:06:07.478070Z  INFO servers::promql: row: [Timestamp(Timestamp { value: 100000, unit: Millisecond }), Float64(OrderedFloat(49.0))]
  ```
- mocked return value
  image

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#596

### CODE REVIEW

**Comment by waynexia** [comment]
I'm also wondering if it's beneficial to provide other ways to perform promql queries, like via SDK or special SQL.

**Comment by killme2008** [comment]
> For the name of this new service, I would prefer to rename it to prometheus as it's a subset of [prometheus HTTP API]> > > (https://prometheus.io/docs/prometheus/latest/querying/api/), and can be extended to implement other methods that are > > not included in this PR. So I propose to rename the current prometheus service to prom_store, which implements > > > Prometheus' remote storage APIs, and rename the new service from promql to prometheus. What do you think about this @killme2008

What about `/v1/prometheus/query`?I think all `prometheus` related APIs must be under `/prometheus` path.

File: src/servers/src/promql.rs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            }
+            Ok(Output::AffectedRows(_)) => Self::error(
+                "unexpected result",
+                "expected data result, but got affected rows",
+            ),
+            Err(err) => Self::error(err.status_code().to_string(), err.to_string()),
+        }
+    }
+
+    fn record_batches_to_data(batches: RecordBatches) -> Result<PromqlData> {
```
Comment (MichaelScofield) [inline src/servers/src/promql.rs:202]: better add a todo here

File: src/servers/src/promql.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    /// Convert from `Result<Output>`
+    pub async fn from_query_result(result: Result<Output>) -> Json<Self> {
+        match result {
+            Ok(Output::RecordBatches(batches)) => {
+                Self::success(Self::record_batches_to_data(batches).unwrap())
+            }
+            Ok(Output::Stream(stream)) => {
+                let record_batches = RecordBatches::try_collect(stream).await.unwrap();
```
Comment (MichaelScofield) [inline src/servers/src/promql.rs:191]: Is it safe to unwrap here?

**Comment by MichaelScofield** [comment]
Seems promql is only executed via http? In that case, I suggest not using `PromqlServer` at all, integrate `PromqlHandler` into struct `HttpServer`, just like other protocol handlers. 

By the way, I think we can get rid of `PromqlHandler` as well, make executing promql as a method of `SqlQueryHandler`, then rename `SqlQueryHandler` to `QueryLangHandler` (maybe other better names). Because promql is translated to logical plan to be actually executed, I think we can unify sql and promql execution path somehow. I've came up with similar thoughts in #766 , if we could make standalone and distribute instance both execute logical plan directly. In that way, promql could be executed in distribute mode, too.

**Comment by waynexia** [comment]
After discussing with @killme2008, we think it's better to provide promql query through two ways:
- via prometheus' official [HTTP API](https://prometheus.io/docs/prometheus/latest/querying/api/), which is partially implemented in this PR
- via our HTTP services, using a new path `/v1/promql`, similar to `/v1/sql`

Implement prometheus' official API can make GreptimeDB a drop-in replacement for Prometheus. Existing systems like grafana only need to change the server address/port to migrate. While`/v1/promql` is providing the promql ability to query GreptimeDB, which is more lightweight to use and flexible to implement.

**Comment by waynexia** [comment]
>By the way, I think we can get rid of PromqlHandler as well, make executing promql as a method of SqlQueryHandler, then rename SqlQueryHandler to QueryLangHandler (maybe other better names). Because promql is translated to logical plan to be actually executed, I think we can unify sql and promql execution path somehow. I've came up with similar thoughts in https://github.com/GreptimeTeam/greptimedb/issues/766 , if we could make standalone and distribute instance both execute logical plan directly. In that way, promql could be executed in distribute mode, too.

Looking great 👍 

>Seems promql is only executed via http? In that case, I suggest not using PromqlServer at all, integrate PromqlHandler into struct HttpServer, just like other protocol handlers.

Consider implementing the prometheus' HTTP API (described in https://github.com/GreptimeTeam/greptimedb/pull/924#issuecomment-1413098309), is it still necessary to add `PromqlServer` to implement prometheus' API?

**Comment by MichaelScofield** [comment]
@waynexia If we are to make our db as a drop in replacement for prometheus, a dedicated server(then a dedicated port) is better. So that prometheus queries do not mix-up with other protocols API pathes.

File: src/servers/src/promql.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+use tower::ServiceBuilder;
+use tower_http::auth::AsyncRequireAuthorizationLayer;
+use tower_http::trace::TraceLayer;
+
+use crate::auth::UserProviderRef;
+use crate::error::{AlreadyStartedSnafu, Result, StartHttpSnafu};
+use crate::http::authorize::HttpAuth;
+use crate::server::Server;
+
+pub const PROMQL_API_VERSION: &str = "v1";
```
[the remark this answers is not in the document]
Comment (waynexia) [reply src/servers/src/promql.rs:42]: This is from prometheus' HTTP API version. We are simulating as a prometheus server so this part needs to be the same.

**Comment by waynexia** [comment]
I'd file two follow-up PRs for (1) rename `prometheus` service to `prom_store` and the new dedicated service `promql` to `prometheus` (I found too many existing references to include in this PR), and (2) refactor `SqlQueryHandler` trait and error types between `servers` mod and `query` mod.

File: src/frontend/src/instance.rs [hunk 4]
```diff
@@ -460,6 +466,18 @@ impl SqlQueryHandler for Instance {
...
 
+    async fn do_promql_query(&self, query: &str, _: QueryContextRef) -> Vec<Result<Output>> {
+        if let Some(handler) = &self.promql_handler {
+            let result = handler
+                .do_query(query)
+                .await
+                .context(ExecutePromqlSnafu { query });
+            vec![result]
+        } else {
+            vec![]
```
Comment (killme2008) [inline src/frontend/src/instance.rs:477]: Return an unsupported error would be better.

File: src/frontend/src/promql.rs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PromqlOptions {
+    pub enable: bool,
```
Comment (killme2008) [inline src/frontend/src/promql.rs:19]: `enable` is unnecessary.If the `addr` is provided, then we enable it by default.

File: src/servers/src/promql.rs [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    /// TODO(ruihang): implement this conversion method
+    fn record_batches_to_data(batches: RecordBatches) -> Result<PromqlData> {
+        debug!("schema: {:?}", batches.schema());
+        for batch in batches.iter() {
+            for row in batch.rows() {
+                debug!("row: {row:?}",);
+            }
+        }
```
Comment (killme2008) [inline src/servers/src/promql.rs:216]: remove these debug code?

**Comment by MichaelScofield** [comment]
There're conflicts. Also would you kind enough to create an issue and assign to me, that our gRPC SDK needs to support promql? thx

**Comment by waynexia** [comment]
> There're conflicts. Also would you kind enough to create an issue and assign to me, that our gRPC SDK needs to support promql? thx

Opened #936 for this

File: src/servers/src/promql.rs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    pub fn set_user_provider(&mut self, user_provider: UserProviderRef) {
+        debug_assert!(self.user_provider.is_none());
+        self.user_provider = Some(user_provider);
+    }
+
+    pub fn make_app(&self) -> Router {
+        // TODO(ruihang): implement format_query, series, labels, values, query_examplars and targets methods
+        let router = Router::new()
```
[the remark this answers is not in the document]
Comment (waynexia) [reply src/servers/src/promql.rs:73]: Not sure if [2fc0fd9](https://github.com/GreptimeTeam/greptimedb/pull/924/commits/2fc0fd93405a97dcb4fe998ded01527811ade7df) is the expected. This service only has one layer of API, is it not that "nest"? I follow the [example](https://docs.rs/axum/latest/axum/routing/struct.Router.html#example-1) of `Router`'s document.

**Review by MichaelScofield — APPROVED** [review]

File: src/servers/src/promql.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (killme2008) [reply src/servers/src/promql.rs:74]: I meant to use the nested route to avoid `format!("/{PROMQL_API_VERSION}/API` repeatedly.

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by killme2008** [comment]
@waynexia  Test failure `integration_http_file_test::test_promql_api`

**Comment by waynexia** [comment]
> @waynexia Test failure `integration_http_file_test::test_promql_api`

Fixed in c67d85a. I only checked `/v1/query` with postman ....

**Pull request merged**